File: src/lib/signals.ts
// The signal sources behind My Wellness.
//
// A "signal" is anything Aurelia can read to decide what a session should be.
// They are grouped by what they tell it about: the body, the day, and the room
// you are in. A real build swaps this for the `integrations` table and an OAuth
// connection per source — the shape here is deliberately close to it.

import type { LucideIcon } from 'lucide-react'
import { CalendarDays, CircleDot, CloudSun, MapPin, MessagesSquare, Watch } from 'lucide-react'

export type SignalGroup = 'Biological' | 'Cognitive' | 'Atmospheric'

export interface SignalSource {
  id: string
  name: string
  group: SignalGroup
  icon: LucideIcon
  /** Whether it is connected out of the box, before the visitor touches anything. */
  defaultOn: boolean
  /** What Aurelia actually reads from it — shown on the row, so consent is informed. */
  reads: string
}

export const SIGNAL_GROUPS: SignalGroup[] = ['Biological', 'Cognitive', 'Atmospheric']

export const SIGNAL_SOURCES: SignalSource[] = [
  {
    id: 'apple-watch',
    name: 'Apple Watch',
    group: 'Biological',
    icon: Watch,
    defaultOn: true,
    reads: 'Heart rate, sleep stages, activity',
  },
  {
    id: 'oura-ring',
    name: 'Oura Ring',
    group: 'Biological',
    icon: CircleDot,
    defaultOn: false,
    reads: 'Readiness, HRV, body temperature',
  },
  {
    id: 'google-calendar',
    name: 'Google Calendar',
    group: 'Cognitive',
    icon: CalendarDays,
    defaultOn: true,
    reads: 'Busy blocks only — never titles or guests',
  },
  {
    id: 'conversation-history',
    name: 'Conversation History',
    group: 'Cognitive',
    icon: MessagesSquare,
    defaultOn: true,
    reads: 'What you asked Aurelia for, and what helped',
  },
  {
    id: 'weather-data',
    name: 'Weather Data',
    group: 'Atmospheric',
    icon: CloudSun,
    defaultOn: true,
    reads: 'Daylight, pressure, temperature',
  },
  {
    id: 'location-data',
    name: 'Location Data',
    group: 'Atmospheric',
    icon: MapPin,
    defaultOn: false,
    reads: 'Coarse area, to know the season and daylight',
  },
]

export function sourcesInGroup(group: SignalGroup) {
  return SIGNAL_SOURCES.filter((source) => source.group === group)
}

/** The state the screen opens in. */
export function defaultConnections(): Record<string, boolean> {
  return Object.fromEntries(SIGNAL_SOURCES.map((source) => [source.id, source.defaultOn]))
}

File: lib/core/data/signals.dart
/// The signal sources behind My Wellness — the mobile mirror of the web app's
/// `src/lib/signals.ts`.
///
/// A "signal" is anything Aurelia can read to decide what a session should be.
/// They are grouped by what they tell it about: the body, the day, and the room
/// you are in. A real build swaps this for the `integrations` table and an OAuth
/// connection per source.
library;

import 'package:flutter/material.dart';

enum SignalGroup { biological, cognitive, atmospheric }

extension SignalGroupLabel on SignalGroup {
  String get label => switch (this) {
        SignalGroup.biological => 'Biological',
        SignalGroup.cognitive => 'Cognitive',
        SignalGroup.atmospheric => 'Atmospheric',
      };
}

class SignalSource {
  const SignalSource({
    required this.id,
    required this.name,
    required this.group,
    required this.icon,
    required this.defaultOn,
    required this.reads,
  });

  final String id;
  final String name;
  final SignalGroup group;
  final IconData icon;

  /// Whether it is connected out of the box, before the visitor touches it.
  final bool defaultOn;

  /// What Aurelia actually reads from it.
  ///
  /// Not rendered — the design's rows are a single line. Kept because it is the
  /// content the consent story needs somewhere, and because whatever the
  /// integration pulls has to match it.
  final String reads;
}

const kSignalSources = <SignalSource>[
  SignalSource(
    id: 'apple-watch',
    name: 'Apple Watch',
    group: SignalGroup.biological,
    icon: Icons.watch_outlined,
    defaultOn: true,
    reads: 'Heart rate, sleep stages, activity',
  ),
  SignalSource(
    id: 'oura-ring',
    name: 'Oura Ring',
    group: SignalGroup.biological,
    icon: Icons.radio_button_checked,
    defaultOn: false,
    reads: 'Readiness, HRV, body temperature',
  ),
  SignalSource(
    id: 'google-calendar',
    name: 'Google Calendar',
    group: SignalGroup.cognitive,
    icon: Icons.calendar_month_outlined,
    defaultOn: true,
    reads: 'Busy blocks only — never titles or guests',
  ),
  SignalSource(
    id: 'conversation-history',
    name: 'Conversation History',
    group: SignalGroup.cognitive,
    icon: Icons.forum_outlined,
    defaultOn: true,
    reads: 'What you asked Aurelia for, and what helped',
  ),
  SignalSource(
    id: 'weather-data',
    name: 'Weather Data',
    group: SignalGroup.atmospheric,
    icon: Icons.wb_cloudy_outlined,
    defaultOn: true,
    reads: 'Daylight, pressure, temperature',
  ),
  SignalSource(
    id: 'location-data',
    name: 'Location Data',
    group: SignalGroup.atmospheric,
    icon: Icons.place_outlined,
    defaultOn: false,
    reads: 'Coarse area, to know the season and daylight',
  ),
];

List<SignalSource> sourcesInGroup(SignalGroup group) =>
    kSignalSources.where((source) => source.group == group).toList();

/// The state the screen opens in.
Map<String, bool> defaultConnections() => {
      for (final source in kSignalSources) source.id: source.defaultOn,
    };
